In [851]:
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)

from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier, VotingClassifier, AdaBoostClassifier
from sklearn.ensemble import BaggingClassifier


import pandas as pd
import numpy as np
import operator
import os

import numpy as np
import scipy as sp
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.decomposition import PCA
from sklearn import preprocessing 

os.environ["KERAS_BACKEND"] = "plaidml.keras.backend"

from sklearn.preprocessing import StandardScaler

from sklearn.pipeline import make_pipeline
from xgboost import XGBClassifier

from sklearn.model_selection import train_test_split
#from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import StandardScaler, RobustScaler, MinMaxScaler

from collections import Counter
from sklearn.metrics import confusion_matrix,classification_report, f1_score
from sklearn.model_selection import cross_val_score
import warnings
import time

import matplotlib.pyplot as plt
from sklearn import svm

from sklearn.feature_selection import VarianceThreshold

from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier, VotingClassifier, AdaBoostClassifier
from sklearn.ensemble import BaggingClassifier

from mlxtend.classifier import StackingClassifier

from sklearn.multiclass import OneVsRestClassifier
from imblearn.combine import SMOTEENN
from imblearn.over_sampling import SMOTE
from imblearn.combine import SMOTETomek


import re
from time import time
import warnings
import joblib
warnings.filterwarnings("ignore")
from sklearn.model_selection import train_test_split

from sklearn.model_selection import StratifiedShuffleSplit, GridSearchCV
from sklearn.metrics import accuracy_score, log_loss, precision_score, recall_score
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis
from sklearn.linear_model import LogisticRegression

pd.set_option('max_colwidth', 1000)




import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import preprocessing
%matplotlib inline

Using plaidml.keras.backend backend.


In [577]:
data = pd.read_csv('/Volumes/MyDisk/Spring2020/Data Mining/deepsolar_tract.csv', encoding='latin1' )
data.head()

Unnamed: 0  tile_count  solar_system_count  total_panel_area         fips  \
0           0         0.0                 0.0          0.000000  27145011200   
1           1        25.0                21.0       1133.436461  27145011301   
2           2         3.0                 3.0         64.505776  27145011302   
3           3         0.0                 0.0          0.000000  27145011304   
4           4         5.0                 5.0        164.583303  27145011400   

   average_household_income          county  education_bachelor  \
0              70352.789869  Stearns County                 569   
1              61727.085202  Stearns County                 674   
2              71496.886583  Stearns County                 854   
3              86840.152755  Stearns County                 640   
4              89135.315597  Stearns County                 654   

   education_college  education_doctoral  education_high_school_graduate  \
0               1690                  13                            1757   
1               1434                 108                             767   
2               1459                  31                            1541   
3               1116                  68                            1095   
4               1314                  15                             982   

   education_less_than_high_school  education_master  education_population  \
0                              336               157                  4548   
1                              222               285                  3533   
2                              289               276                  4480   
3                              231               270                  3456   
4                              163               170                  3348   

   education_professional_school  employed  gini_index  \
0                             26      3569      0.3490   
1                             43      4795      0.4074   
2                             30      3550      0.3926   
3                             36      3483      0.3949   
4                             50      2887      0.4463   

   heating_fuel_coal_coke  heating_fuel_electricity  \
0                     200                       448   
1                      20                       379   
2                      69                       440   
3                     188                       442   
4                      96                       497   

   heating_fuel_fuel_oil_kerosene  heating_fuel_gas  \
0                              98              1720   
1                              17              1799   
2                              45              2098   
3                              61              1113   
4                              47              1202   

   heating_fuel_housing_unit_count  heating_fuel_none  heating_fuel_other  \
0                             2527                 10                  51   
1                             2230                  0                  15   
2                             2698                 29                  17   
3                             1833                 13                  16   
4                             1917                 31                  44   

   heating_fuel_solar  land_area  per_capita_income  population  \
0                   0  147.73120            26999.0        6577   
1                   0   16.61679            20951.0        8020   
2                   0   37.44482            28021.0        6971   
3                   0   75.21147            29275.0        6328   
4                   0   44.00151            32770.0        5264   

   population_density  poverty_family_below_poverty_level  \
0            44.52005                                  85   
1           482.64430                                 189   
2           186.16730                                 134   
3            84.13610                                  87   
4          

In [630]:
# Checking the dimension of the data - [rows, columns]

data.shape

(72537, 169)

In [646]:


data.isnull().sum()

Unnamed: 0                                     0
tile_count                                     0
solar_system_count                             0
total_panel_area                               0
fips                                           0
average_household_income                     886
county                                         0
education_bachelor                             0
education_college                              0
education_doctoral                             0
education_high_school_graduate                 0
education_less_than_high_school                0
education_master                               0
education_population                           0
education_professional_school                  0
employed                                       0
gini_index                                   847
heating_fuel_coal_coke                         0
heating_fuel_electricity                       0
heating_fuel_fuel_oil_kerosene                 0
heating_fuel_gas    

In [641]:
# Slicing the dataset to reflect the State of California 

# And the checking the sliced data

cali_data = data[data['state']=='ca']
cali_data.shape

(8056, 169)

In [633]:
# Looking into the total number of missing values each variable has

cali_data.isnull().sum()

Unnamed: 0                                   0
tile_count                                   0
solar_system_count                           0
total_panel_area                             0
fips                                         0
average_household_income                    84
county                                       0
education_bachelor                           0
education_college                            0
education_doctoral                           0
education_high_school_graduate               0
education_less_than_high_school              0
education_master                             0
education_population                         0
education_professional_school                0
employed                                     0
gini_index                                  81
heating_fuel_coal_coke                       0
heating_fuel_electricity                     0
heating_fuel_fuel_oil_kerosene               0
heating_fuel_gas                             0
heating_fuel_

## Census Tract Level Environmental and Socioeconomic factors


### 1. Statewide Incentive parameters and Average Electricity Retail Rate

In [618]:
incentive_parameters = [

                        'incentive_count_residential',
                        'net_metering',
                        'feedin_tariff',
                        'cooperate_tax',
                        'property_tax',
                        'sales_tax',
                        'rebate',
                        'avg_electricity_retail_rate',
                        'electricity_price_residential',                                     

                       ]
len(incentive_parameters)

9

In [619]:
# Slicing the demographic variables and reviewing the first ten elements

cali_inc_data = cali_data[incentive_parameters]
cali_inc_data.head(10)

incentive_count_residential  net_metering  feedin_tariff  cooperate_tax  \
108                           75            21              9              0   
109                           75            21              9              0   
110                           75            21              9              0   
111                           75            21              9              0   
112                           75            21              9              0   
113                           75            21              9              0   
114                           75            21              9              0   
115                           75            21              9              0   
116                           75            21              9              0   
117                           75            21              9              0   

     property_tax  sales_tax  rebate  avg_electricity_retail_rate  \
108            38          0      10                        14.72   
109            38          0      10                        14.72   
110            38          0      10                        14.72   
111            38          0      10                        14.72   
112            38          0      10                        14.72   
113            38          0      10                        14.72   
114            38          0      10                        14.72   
115            38          0      10                        14.72   
116            38          0      10                        14.72   
117            38          0      10                        14.72   

     electricity_price_residential  
108                          16.99  
109                          16.99  
110                          16.99  
111                          16.99  
112                          16.99  
113                          16.99  
114                          16.99  
115                          16.99  
116                          16.99  
117                          16.99

In [628]:
# Looking into missing values of California demographic variables

cali_inc_data.isnull().sum()

incentive_count_residential      0
net_metering                     0
feedin_tariff                    0
cooperate_tax                    0
property_tax                     0
sales_tax                        0
rebate                           0
avg_electricity_retail_rate      0
electricity_price_residential    0
dtype: int64

    ** It is shown in the above that the incentive variables for the state of california does not have any missing values 

In [629]:
# For convenince, the clean incentive data slice is labeled as clean

clean_cali_inc_data = cali_inc_data.copy()
clean_cali_inc_data

incentive_count_residential  net_metering  feedin_tariff  \
108                             75            21              9   
109                             75            21              9   
110                             75            21              9   
111                             75            21              9   
112                             75            21              9   
...                            ...           ...            ...   
70868                           75            21              9   
70869                           75            21              9   
70870                           75            21              9   
70871                           75            21              9   
71298                           75            21              9   

       cooperate_tax  property_tax  sales_tax  rebate  \
108                0            38          0      10   
109                0            38          0      10   
110                0            38          0      10   
111                0            38          0      10   
112                0            38          0      10   
...              ...           ...        ...     ...   
70868              0            38          0      10   
70869              0            38          0      10   
70870              0            38          0      10   
70871              0            38          0      10   
71298              0            38          0      10   

       avg_electricity_retail_rate  electricity_price_residential  
108                          14.72                          16.99  
109                          14.72                          16.99  
110                          14.72                          16.99  
111                          14.72                          16.99  
112                          14.72                          16.99  
...                            ...                            ...  
70868                        14.72                          16.99  
70869                        14.72                          16.99  
70870                        14.72                          16.99  
70871                        14.72                          16.99  
71298                        14.72                          16.99  

[8056 rows x 9 columns]

     ** At this stage of the implementation, the California incentive variables do not have missing values
     
     ** Clean California incentive data
     

### 2. Demographic Parameters 

In [587]:
demographic_parameters = [
    'county',
    'average_household_income',
    'gini_index',
    'per_capita_income',                                                    
    'population_density',                                                   
    'education_less_than_high_school_rate',         
'education_high_school_graduate_rate',         
'education_college_rate',                       
'education_bachelor_rate',                      
'education_master_rate',                        
'education_professional_school_rate',           
'education_doctoral_rate',                      
'race_white_rate',                              
'race_black_africa_rate',                       
'race_indian_alaska_rate',                      
'race_asian_rate',                              
'race_islander_rate',                          
'race_other_rate',                              
'race_two_more_rate',
'poverty_family_below_poverty_level_rate',      
'heating_fuel_gas_rate',                        
'heating_fuel_electricity_rate',                
'heating_fuel_fuel_oil_kerosene_rate',          
'heating_fuel_coal_coke_rate',                  
'heating_fuel_solar_rate',                      
'heating_fuel_other_rate',                      
'heating_fuel_none_rate',                       
'household_count',                                
'average_household_size',                      
'housing_unit_count',                             
'housing_unit_occupied_count',                    
'housing_unit_median_value',                   
'housing_unit_median_gross_rent',
'age_18_24_rate',                               
'age_25_34_rate',                               
'age_more_than_85_rate',                        
'age_75_84_rate',                               
'age_35_44_rate',                               
'age_45_54_rate',                               
'age_65_74_rate',                               
'age_55_64_rate',                               
'age_10_14_rate',                               
'age_15_17_rate',                               
'age_5_9_rate',                                 
'household_type_family_rate',                   
'dropout_16_19_inschool_rate',                 
'occupation_construction_rate',                 
'occupation_public_rate',                       
'occupation_information_rate',                  
'occupation_finance_rate',                      
'occupation_education_rate',                    
'occupation_administrative_rate',               
'occupation_manufacturing_rate',                
'occupation_wholesale_rate',                    
'occupation_retail_rate',                       
'occupation_transportation_rate',               
'occupation_arts_rate',                        
'occupation_agriculture_rate',                  
'occupancy_vacant_rate',                        
'occupancy_owner_rate',                         
'mortgage_with_rate',                          
'transportation_home_rate',                     
'transportation_car_alone_rate',                
'transportation_walk_rate',                     
'transportation_carpool_rate',                  
'transportation_motorcycle_rate',               
'transportation_bicycle_rate',                  
'transportation_public_rate',                   
'travel_time_less_than_10_rate',               
'travel_time_10_19_rate',                       
'travel_time_20_29_rate',                       
'travel_time_30_39_rate',                       
'travel_time_40_59_rate',                       
'travel_time_60_89_rate',                       
'health_insurance_public_rate',                 
'health_insurance_none_rate',                                 
]

len(demographic_parameters)

76

In [611]:
# Slicing the demographic variables and reviewing the first ten elements

cali_demo_data = cali_data[demographic_parameters]
cali_demo_data.head(10)

county  average_household_income  gini_index  \
108  Los Angeles County              88640.331010      0.4576   
109  Los Angeles County              99115.358593      0.4084   
110  Los Angeles County             414706.238185      0.5052   
111  Los Angeles County             107896.201580      0.5405   
112  Los Angeles County             123826.462548      0.5640   
113  Los Angeles County              50569.517885      0.4330   
114  Los Angeles County             130238.329911      0.5376   
115  Los Angeles County             102484.424449      0.4666   
116  Los Angeles County              62724.734694      0.4400   
117  Los Angeles County             198889.278901      0.5685   

     per_capita_income  population_density  \
108            19648.0           11309.820   
109            29555.0            8865.547   
110           124712.0            2624.595   
111            47022.0           11027.470   
112            67095.0           12906.290   
113            16951.0           12753.500   
114            51988.0            9255.900   
115            46837.0           19410.790   
116            18892.0            8364.018   
117            95105.0           10101.790   

     education_less_than_high_school_rate  \
108                              0.047858   
109                              0.102294   
110                              0.064271   
111                              0.066475   
112                              0.026752   
113                              0.255090   
114                              0.036670   
115                              0.042698   
116                              0.244304   
117                              0.018913   

     education_high_school_graduate_rate  education_college_rate  \
108                             0.225010                0.407565   
109                             0.203658                0.389957   
110                             0.149701                0.118164   
111                             0.269809                0.218769   
112                             0.119819                0.270535   
113                             0.337725                0.282236   
114                             0.097910                0.263293   
115                             0.113861                0.214109   
116                             0.290823                0.230380   
117                             0.043499                0.164775   

     education_bachelor_rate  education_master_rate  \
108                 0.146662               0.130838   
109                 0.162430               0.114693   
110                 0.289820               0.114970   
111                 0.251698               0.078411   
112                 0.340995               0.122080   
113                 0.063074               0.041118   
114                 0.320865               0.161349   
115                 0.340140               0.144802   
116                 0.179430               0.054114   
117                 0.348936               0.196217   

     education_professional_school_rate  education_doctoral_rate  \
108                            0.020455                 0.021613   
109                            0.013639                 0.013329   
110                            0.211178                 0.051896   
111                            0.058448                 0.056390   
112                            0.087792                 0.032027   
113                            0.000000                 0.020758   
114                            0.087642                 0.032270   
115                            0.108292                 0.036097   
116                            0.000949                 0.000000   
117                            0.154610                 0.073050   

     race_white_rate  race_black_africa_rate  race_indian_alaska_rate  \
108         0.658069                0.013558                 0.003803   
109         0.652914                0.030659    

In [589]:
# Looking into missing values of California demographic variables

cali_demo_data.isnull().sum()

county                                       0
average_household_income                    84
gini_index                                  81
per_capita_income                           50
population_density                          22
education_less_than_high_school_rate        48
education_high_school_graduate_rate         48
education_college_rate                      48
education_bachelor_rate                     48
education_master_rate                       48
education_professional_school_rate          48
education_doctoral_rate                     48
race_white_rate                             45
race_black_africa_rate                      45
race_indian_alaska_rate                     45
race_asian_rate                             45
race_islander_rate                          45
race_other_rate                             45
race_two_more_rate                          45
poverty_family_below_poverty_level_rate     78
heating_fuel_gas_rate                       73
heating_fuel_

    ** All the missing values of demographic variables in the State of California are listed above
    
    ** Considering the number of samples we have, over 8,000 samples, the number of missing values are manageable
    
    ** Hence these missing values will be filled with some values of the corresponding variable. 
    
    ** To attain the balance of the variable values, filling the missing values with the mean looks reasonable.
    
    ** Hence, the mean values of respective counties are used to fill this values. 

In [590]:
# Extracting the demographic variables with missing values

demo_var = cali_demo_data.isnull().sum() > 0
mis_val_var = demo_var[demo_var == True].keys()
mis_val_var

Index(['average_household_income', 'gini_index', 'per_capita_income',
       'population_density', 'education_less_than_high_school_rate',
       'education_high_school_graduate_rate', 'education_college_rate',
       'education_bachelor_rate', 'education_master_rate',
       'education_professional_school_rate', 'education_doctoral_rate',
       'race_white_rate', 'race_black_africa_rate', 'race_indian_alaska_rate',
       'race_asian_rate', 'race_islander_rate', 'race_other_rate',
       'race_two_more_rate', 'poverty_family_below_poverty_level_rate',
       'heating_fuel_gas_rate', 'heating_fuel_electricity_rate',
       'heating_fuel_fuel_oil_kerosene_rate', 'heating_fuel_coal_coke_rate',
       'heating_fuel_solar_rate', 'heating_fuel_other_rate',
       'heating_fuel_none_rate', 'average_household_size',
       'housing_unit_median_value', 'housing_unit_median_gross_rent',
       'age_18_24_rate', 'age_25_34_rate', 'age_more_than_85_rate',
       'age_75_84_rate', 'age_35_44_rate

In [591]:
#  Extracting State of California counties and checking them with missing demographic variable values

cali_county = cali_demo_data['county'].unique().tolist()
cali_county

['Los Angeles County',
 'San Mateo County',
 'San Joaquin County',
 'San Diego County',
 'Tuolumne County',
 'Marin County',
 'Riverside County',
 'San Francisco County',
 'Contra Costa County',
 'Butte County',
 'Orange County',
 'San Luis Obispo County',
 'San Benito County',
 'Alameda County',
 'Kern County',
 'Napa County',
 'Santa Clara County',
 'Tulare County',
 'Humboldt County',
 'Placer County',
 'Sonoma County',
 'Fresno County',
 'Monterey County',
 'Plumas County',
 'Merced County',
 'Stanislaus County',
 'Calaveras County',
 'Santa Cruz County',
 'Yuba County',
 'San Bernardino County',
 'Santa Barbara County',
 'Solano County',
 'Sacramento County',
 'Madera County',
 'Tehama County',
 'Lake County',
 'Imperial County',
 'Del Norte County',
 'Kings County',
 'Inyo County',
 'Lassen County',
 'Ventura County',
 'Nevada County',
 'Mendocino County',
 'Modoc County',
 'Colusa County',
 'El Dorado County',
 'Sutter County',
 'Alpine County',
 'Yolo County',
 'Siskiyou County

    ** In the above sections, all the demographic variables with missing values are listed,
    
    ** Moreover, all the counties found in california are listed above,
    
    ** As shown below, a function is used fill the values with the mean values of corresponding county, 

In [592]:
# Function filling missing values of California environmental variables based on the mean of respective counties

def fill_miss_data(dataset, mis_val_var, mis_val_county):
    
    # Iterate over the list of counties with missing values of environmental variables
    for i in range(len(mis_val_county)):
        
        # Slicing the dataset to a county level
        data = dataset[dataset['county'] == mis_val_county[i]]
        
        #Iterating over the missing variables with in the sliced county
        for j in range(len(mis_val_var)):
            
            var = mis_val_var[j]
            mean = data[var].mean()              # Computation of the mode value of an envt variable
            dataset[var].fillna(mean, inplace=True) # Filling missing values with mode values of respective county
            
    return dataset

In [593]:
# Appyiing the function to fill the values and checking for missing values again

cali_demo_data = fill_miss_data(cali_demo_data, mis_val_var, cali_county)
cali_demo_data.isnull().sum()

county                                     0
average_household_income                   0
gini_index                                 0
per_capita_income                          0
population_density                         0
education_less_than_high_school_rate       0
education_high_school_graduate_rate        0
education_college_rate                     0
education_bachelor_rate                    0
education_master_rate                      0
education_professional_school_rate         0
education_doctoral_rate                    0
race_white_rate                            0
race_black_africa_rate                     0
race_indian_alaska_rate                    0
race_asian_rate                            0
race_islander_rate                         0
race_other_rate                            0
race_two_more_rate                         0
poverty_family_below_poverty_level_rate    0
heating_fuel_gas_rate                      0
heating_fuel_electricity_rate              0
heating_fu

In [594]:
# Since the demographic variables do not have missing values anymore, it is labeled as clean for convenience

clean_cali_demo_data = cali_demo_data.copy()
clean_cali_demo_data.head()

county  average_household_income  gini_index  \
108  Los Angeles County              88640.331010      0.4576   
109  Los Angeles County              99115.358593      0.4084   
110  Los Angeles County             414706.238185      0.5052   
111  Los Angeles County             107896.201580      0.5405   
112  Los Angeles County             123826.462548      0.5640   

     per_capita_income  population_density  \
108            19648.0           11309.820   
109            29555.0            8865.547   
110           124712.0            2624.595   
111            47022.0           11027.470   
112            67095.0           12906.290   

     education_less_than_high_school_rate  \
108                              0.047858   
109                              0.102294   
110                              0.064271   
111                              0.066475   
112                              0.026752   

     education_high_school_graduate_rate  education_college_rate  \
108                             0.225010                0.407565   
109                             0.203658                0.389957   
110                             0.149701                0.118164   
111                             0.269809                0.218769   
112                             0.119819                0.270535   

     education_bachelor_rate  education_master_rate  \
108                 0.146662               0.130838   
109                 0.162430               0.114693   
110                 0.289820               0.114970   
111                 0.251698               0.078411   
112                 0.340995               0.122080   

     education_professional_school_rate  education_doctoral_rate  \
108                            0.020455                 0.021613   
109                            0.013639                 0.013329   
110                            0.211178                 0.051896   
111                            0.058448                 0.056390   
112                            0.087792                 0.032027   

     race_white_rate  race_black_africa_rate  race_indian_alaska_rate  \
108         0.658069                0.013558                 0.003803   
109         0.652914                0.030659                 0.000000   
110         0.853829                0.010620                 0.000000   
111         0.814982                0.016468                 0.005651   
112         0.801053                0.005266                 0.002633   

     race_asian_rate  race_islander_rate  race_other_rate  race_two_more_rate  \
108         0.165013            0.003803         0.102679            0.053075   
109         0.124716            0.000000         0.143073            0.048637   
110         0.057015            0.000000         0.022079            0.056456   
111         0.101711            0.000000         0.025509            0.035680   
112         0.138092            0.000000         0.035693            0.017262   

     poverty_family_below_poverty_level_rate  heating_fuel_gas_rate  \
108                                 0.030888               0.788328   
109                                 0.060877               0.891746   
110                                 0.018315               0.801512   
111                                 0.147354               0.559985   
112                                 0.054127               0.704210   

     heating_fuel_electricity_rate  heating_fuel_fuel_oil_kerosene_rate  \
108                       0.179443                             0.000000   
109                       0.104195                             0.000000   
110                       0.186200                             0.000000   
111                       0.418578                             0.000000   
112                       0.270093                             0.005467   

     heating_fuel_coal_coke_rate  heating_fuel_solar_rate  \
108                          0.0                 0.011324   
109

     ** At this part of the project, the California demographical variables do not have missing values
     
     ** Clean California demographical variables

### Environmental Condition Parameters

In [623]:
envt_parameters = [
    'county',
'heating_design_temperature',                  
'cooling_design_temperature',                  
'earth_temperature_amplitude',             
'frost_days',                                 
'air_temperature',                             
'relative_humidity',                          
'daily_solar_radiation',                       
'atmospheric_pressure',                        
'wind_speed',                                 
'earth_temperature',                           
'heating_degree_days',                         
'cooling_degree_days',
]
len(envt_parameters)

13

In [624]:
# Slicing the dataset to take closet look at the envornmental variables/parameters

cali_envt_data = cali_data[envt_parameters]
cali_envt_data.head(10)

county  heating_design_temperature  \
108  Los Angeles County                        7.39   
109  Los Angeles County                        7.39   
110  Los Angeles County                        2.35   
111  Los Angeles County                        2.35   
112  Los Angeles County                        2.35   
113  Los Angeles County                        7.39   
114  Los Angeles County                        2.35   
115  Los Angeles County                        2.35   
116  Los Angeles County                        2.35   
117  Los Angeles County                        2.35   

     cooling_design_temperature  earth_temperature_amplitude  frost_days  \
108                       28.26                        13.18         0.0   
109                       28.26                        13.18         0.0   
110                       33.22                        22.20         3.0   
111                       33.22                        22.20         3.0   
112                       33.22                        22.20         3.0   
113                       28.26                        13.18         0.0   
114                       33.22                        22.20         3.0   
115                       33.22                        22.20         3.0   
116                       33.22                        22.20         3.0   
117                       33.22                        22.20         3.0   

     air_temperature  relative_humidity  daily_solar_radiation  \
108             17.6              0.515                   5.20   
109             17.6              0.515                   5.20   
110             17.2              0.401                   5.41   
111             17.2              0.401                   5.41   
112             17.2              0.401                   5.41   
113             17.6              0.515                   5.20   
114             17.2              0.401                   5.41   
115             17.2              0.401                   5.41   
116             17.2              0.401                   5.41   
117             17.2              0.401                   5.41   

     atmospheric_pressure  wind_speed  earth_temperature  heating_degree_days  \
108                  98.6         4.6               19.5                767.0   
109                  98.6         4.6               19.5                767.0   
110                  94.7         4.4               19.8               1208.0   
111                  94.7         4.4               19.8               1208.0   
112                  94.7         4.4               19.8               1208.0   
113                  98.6         4.6               19.5                767.0   
114                  94.7         4.4               19.8               1208.0   
115                  94.7         4.4               19.8               1208.0   
116                  94.7         4.4               19.8               1208.0   
117                  94.7         4.4               19.8               1208.0   

     cooling_degree_days  
108               2841.0  
109               2841.0  
110               2815.0  
111               2815.0  
112               2815.0  
113               2841.0  
114               2815.0  
115               2815.0  
116               2815.0  
117               2815.0

In [625]:
# Checking missing values of California environmental variables

cali_envt_data.isnull().sum() 

county                          0
heating_design_temperature     16
cooling_design_temperature     16
earth_temperature_amplitude    16
frost_days                     16
air_temperature                16
relative_humidity              16
daily_solar_radiation          16
atmospheric_pressure           16
wind_speed                     16
earth_temperature              16
heating_degree_days            16
cooling_degree_days            16
dtype: int64

In [626]:
# The fact they all have the same number of missing values implies they might be missing from the same places

# USING ONE OF THE VARIABLES  to check if the missing values are from the same places

# It is here shown that all the missing values are from the same places of different counties

cali_envt_data[cali_envt_data['wind_speed'].isnull() == True]

county  heating_design_temperature  \
633        Los Angeles County                         NaN   
635        Los Angeles County                         NaN   
4539         San Mateo County                         NaN   
8229         San Diego County                         NaN   
14157            Marin County                         NaN   
17791    San Francisco County                         NaN   
17979     Contra Costa County                         NaN   
19220           Orange County                         NaN   
19966  San Luis Obispo County                         NaN   
23017          Alameda County                         NaN   
27310         Humboldt County                         NaN   
28776           Sonoma County                         NaN   
29519         Monterey County                         NaN   
37688       Santa Cruz County                         NaN   
55344        Del Norte County                         NaN   
60548        Mendocino County                         NaN   

       cooling_design_temperature  earth_temperature_amplitude  frost_days  \
633                           NaN                          NaN         NaN   
635                           NaN                          NaN         NaN   
4539                          NaN                          NaN         NaN   
8229                          NaN                          NaN         NaN   
14157                         NaN                          NaN         NaN   
17791                         NaN                          NaN         NaN   
17979                         NaN                          NaN         NaN   
19220                         NaN                          NaN         NaN   
19966                         NaN                          NaN         NaN   
23017                         NaN                          NaN         NaN   
27310                         NaN                          NaN         NaN   
28776                         NaN                          NaN         NaN   
29519                         NaN                          NaN         NaN   
37688                         NaN                          NaN         NaN   
55344                         NaN                          NaN         NaN   
60548                         NaN                          NaN         NaN   

       air_temperature  relative_humidity  daily_solar_radiation  \
633                NaN                NaN                    NaN   
635                NaN                NaN                    NaN   
4539               NaN                NaN                    NaN   
8229               NaN                NaN                    NaN   
14157              NaN                NaN                    NaN   
17791              NaN                NaN                    NaN   
17979              NaN                NaN                    NaN   
19220              NaN                NaN                    NaN   
19966              NaN                NaN                    NaN   
23017              NaN                NaN                    NaN   
27310              NaN                NaN                    NaN   
28776              NaN                NaN                    NaN   
29519              NaN                NaN                    NaN   
37688              NaN                NaN                    NaN   
55344              NaN                NaN                    NaN   
60548              NaN                NaN                    NaN   

       atmospheric_pressure  wind_speed  earth_temperature  \
633                     NaN         NaN                NaN   
635                     NaN         NaN                NaN   
4539                    NaN         NaN                NaN   
8229                    NaN         NaN                NaN   
14157                   NaN         NaN                NaN   
17791                   NaN         NaN                NaN   
17979                   NaN         NaN                NaN   
19220

In [599]:
# Listing the counties in the State of California with missing environmental variable values

mis_val_county = cali_envt_data[cali_envt_data['wind_speed'].isnull() == True]['county'].unique().tolist()
mis_val_county

['Los Angeles County',
 'San Mateo County',
 'San Diego County',
 'Marin County',
 'San Francisco County',
 'Contra Costa County',
 'Orange County',
 'San Luis Obispo County',
 'Alameda County',
 'Humboldt County',
 'Sonoma County',
 'Monterey County',
 'Santa Cruz County',
 'Del Norte County',
 'Mendocino County']

In [600]:
# These are the environmental variables with missing values in the State of California

mis_val_var = envt_parameters[1:]
mis_val_var

['heating_design_temperature',
 'cooling_design_temperature',
 'earth_temperature_amplitude',
 'frost_days',
 'air_temperature',
 'relative_humidity',
 'daily_solar_radiation',
 'atmospheric_pressure',
 'wind_speed',
 'earth_temperature',
 'heating_degree_days',
 'cooling_degree_days']

    ** As shown in the above , we have 16 missing values from each of those environmental variables listed above,
    
    ** And as shown in the above, these missing values belong to different counties,
    
    ** Hence, the mean values of corresponding county are used to fill the missing values

In [602]:
# Running the previous function to fill the missing values and checking for missing values again

cali_envt_data = fill_miss_data(cali_envt_data, mis_val_var, mis_val_county )
cali_envt_data.isnull().sum()

county                         0
heating_design_temperature     0
cooling_design_temperature     0
earth_temperature_amplitude    0
frost_days                     0
air_temperature                0
relative_humidity              0
daily_solar_radiation          0
atmospheric_pressure           0
wind_speed                     0
earth_temperature              0
heating_degree_days            0
cooling_degree_days            0
dtype: int64

In [603]:
# Finalize and copy the clean environmental variables and labeling it as clean

clean_cali_envt_data = cali_envt_data.copy()
clean_cali_envt_data.head()

county  heating_design_temperature  \
108  Los Angeles County                        7.39   
109  Los Angeles County                        7.39   
110  Los Angeles County                        2.35   
111  Los Angeles County                        2.35   
112  Los Angeles County                        2.35   

     cooling_design_temperature  earth_temperature_amplitude  frost_days  \
108                       28.26                        13.18         0.0   
109                       28.26                        13.18         0.0   
110                       33.22                        22.20         3.0   
111                       33.22                        22.20         3.0   
112                       33.22                        22.20         3.0   

     air_temperature  relative_humidity  daily_solar_radiation  \
108             17.6              0.515                   5.20   
109             17.6              0.515                   5.20   
110             17.2              0.401                   5.41   
111             17.2              0.401                   5.41   
112             17.2              0.401                   5.41   

     atmospheric_pressure  wind_speed  earth_temperature  heating_degree_days  \
108                  98.6         4.6               19.5                767.0   
109                  98.6         4.6               19.5                767.0   
110                  94.7         4.4               19.8               1208.0   
111                  94.7         4.4               19.8               1208.0   
112                  94.7         4.4               19.8               1208.0   

     cooling_degree_days  
108               2841.0  
109               2841.0  
110               2815.0  
111               2815.0  
112               2815.0

     ** At this stage of the implementation, the California environemental variables do not have missing values
     ** Clean California environmental data

### Political Views

In [604]:
pol_parameters = [
    
'voting_2016_dem_percentage',                     
'voting_2016_gop_percentage',                     
'voting_2016_dem_win',                            
'voting_2012_dem_percentage',                 
'voting_2012_gop_percentage',                
'voting_2012_dem_win'                           
]

len(political_parameters)

6

In [605]:
# Looking into the missing values of Poltical view data values in the State of California

# As shown here below, this part of the dataset fortunately does not have missing values

cali_pol_data = cali_data[pol_parameters]
cali_pol_data.isnull().sum()

voting_2016_dem_percentage    0
voting_2016_gop_percentage    0
voting_2016_dem_win           0
voting_2012_dem_percentage    0
voting_2012_gop_percentage    0
voting_2012_dem_win           0
dtype: int64

In [606]:
# However, mapping the True and False value to a corresponding binary variables enhances applying computations

# Mapping True and False values of our attributes into 1 and 0 respectively

vot_map = {True: 1, False: 0}

cali_pol_data['voting_2016_dem_win'] = pol_data['voting_2016_dem_win'].map(vot_map)
cali_pol_data['voting_2012_dem_win'] = pol_data['voting_2016_dem_win'].map(vot_map)
cali_pol_data.head()

voting_2016_dem_percentage  voting_2016_gop_percentage  \
108                    0.714072                    0.233887   
109                    0.714072                    0.233887   
110                    0.714072                    0.233887   
111                    0.714072                    0.233887   
112                    0.714072                    0.233887   

     voting_2016_dem_win  voting_2012_dem_percentage  \
108                    1                       0.689   
109                    1                       0.689   
110                    1                       0.689   
111                    1                       0.689   
112                    1                       0.689   

     voting_2012_gop_percentage  voting_2012_dem_win  
108                       0.288                    1  
109                       0.288                    1  
110                       0.288                    1  
111                       0.288                    1  
112                       0.288                    1

In [607]:
# Finalize and copy the clean political variables and labeling it as clean

clean_cali_pol_data = cali_pol_data.copy()
clean_cali_pol_data

voting_2016_dem_percentage  voting_2016_gop_percentage  \
108                      0.714072                    0.233887   
109                      0.714072                    0.233887   
110                      0.714072                    0.233887   
111                      0.714072                    0.233887   
112                      0.714072                    0.233887   
...                           ...                         ...   
70868                    0.326794                    0.617123   
70869                    0.326794                    0.617123   
70870                    0.326794                    0.617123   
70871                    0.326794                    0.617123   
71298                    0.333149                    0.580931   

       voting_2016_dem_win  voting_2012_dem_percentage  \
108                      1                       0.689   
109                      1                       0.689   
110                      1                       0.689   
111                      1                       0.689   
112                      1                       0.689   
...                    ...                         ...   
70868                    0                       0.358   
70869                    0                       0.358   
70870                    0                       0.358   
70871                    0                       0.358   
71298                    0                       0.366   

       voting_2012_gop_percentage  voting_2012_dem_win  
108                         0.288                    1  
109                         0.288                    1  
110                         0.288                    1  
111                         0.288                    1  
112                         0.288                    1  
...                           ...                  ...  
70868                       0.613                    0  
70869                       0.613                    0  
70870                       0.613                    0  
70871                       0.613                    0  
71298                       0.594                    0  

[8056 rows x 6 columns]

      * The political views variables from the State of California is clean and ready to be processed at this stage

## Merging Clean Environmental and Socio-economic Data Slices

In [834]:
# To perform computations on all the slices combined, all the slices are merged at this section

# Moreover, the dimensions of the dataset are also checked for consistency

slices = [clean_cali_inc_data, clean_cali_demo_data, clean_cali_envt_data, clean_cali_pol_data]
dataset = pd.concat(slices, axis=1, sort=False)
dataset.drop('county', axis=1, inplace=True)
dataset.shape

(8056, 102)

In [835]:
# Further checking for missing values and consistency checking

dataset.isnull().sum()

incentive_count_residential                0
net_metering                               0
feedin_tariff                              0
cooperate_tax                              0
property_tax                               0
sales_tax                                  0
rebate                                     0
avg_electricity_retail_rate                0
electricity_price_residential              0
average_household_income                   0
gini_index                                 0
per_capita_income                          0
population_density                         0
education_less_than_high_school_rate       0
education_high_school_graduate_rate        0
education_college_rate                     0
education_bachelor_rate                    0
education_master_rate                      0
education_professional_school_rate         0
education_doctoral_rate                    0
race_white_rate                            0
race_black_africa_rate                     0
race_india

     ** At this stage of the project, the environmental and socio-economic variables reflecting the State of California are clean and ready for further regression or classification processing

# Classification

###     A.   Feature Selection 

      ** Until this stage, we have been cleaning socio-economic dataset of the State of California 
      
      ** The State of California's socio-economic dataset contains 8056 data records with 104 dimensions
      
      ** Feature selection is applied at this stage to drop features with little contribution for classification 

In [836]:
# Removing features with zero standard deviation - having values that does not change 

print("Original dataset shape - ", dataset.shape)

# Remove Constant Features
constant_features = [feat for feat in dataset.columns if (dataset[feat].std() == 0 or dataset[feat].min() == dataset[feat].max())]

print('\nThe following features are constant with little contribution for classification, Hence they are dropped \n\n', constant_features)
print('\n')

# Dropping all the data with constant values 

dataset.drop(labels=constant_features, axis=1, inplace=True)
print("Reduced feature dataset shape -",dataset.shape)

Original dataset shape -  (8056, 102)

The following features are constant with little contribution for classification, Hence they are dropped 

 ['incentive_count_residential', 'net_metering', 'feedin_tariff', 'cooperate_tax', 'property_tax', 'sales_tax', 'rebate', 'avg_electricity_retail_rate', 'electricity_price_residential']


Reduced feature dataset shape - (8056, 93)


### B. Feature Creation

In [747]:
# Looking into the distribution of the data

dataset.describe()

average_household_income   gini_index  per_capita_income  \
count               8056.000000  8056.000000        8056.000000   
mean               85433.476192     0.420459       31141.920827   
std                43743.271886     0.059864       18650.855284   
min                 9040.000000     0.012800         470.000000   
25%                55694.517065     0.380400       17799.000000   
50%                75782.095234     0.416500       26924.000000   
75%               103458.331336     0.454100       39263.500000   
max               537130.215827     0.731700      181215.000000   

       population_density  education_less_than_high_school_rate  \
count         8056.000000                           8056.000000   
mean          8600.291584                              0.190524   
std           9651.212644                              0.159264   
min              0.000000                              0.000000   
25%           2669.763750                              0.062448   
50%           6315.999500                              0.140160   
75%          11045.155000                              0.283611   
max         173333.700000                              1.000000   

       education_high_school_graduate_rate  education_college_rate  \
count                          8056.000000             8056.000000   
mean                              0.208683                0.294265   
std                               0.080379                0.090287   
min                               0.000000                0.000000   
25%                               0.152630                0.231739   
50%                               0.213870                0.295907   
75%                               0.266950                0.357706   
max                               0.522634                1.000000   

       education_bachelor_rate  education_master_rate  \
count              8056.000000            8056.000000   
mean                  0.193033               0.075635   
std                   0.113595               0.066832   
min                   0.000000               0.000000   
25%                   0.096597               0.023827   
50%                   0.180932               0.056847   
75%                   0.280578               0.111044   
max                   1.000000               1.000000   

       education_professional_school_rate  education_doctoral_rate  \
count                         8056.000000              8056.000000   
mean                             0.022714                 0.015146   
std                              0.027514                 0.022173   
min                              0.000000                 0.000000   
25%                              0.004127                 0.001121   
50%                              0.013443                 0.007500   
75%                              0.030362                 0.020258   
max                              0.235900                 0.274282   

       race_white_rate  race_black_africa_rate  race_indian_alaska_rate  \
count      8056.000000             8056.000000              8056.000000   
mean          0.623674                0.059558                 0.007771   
std           0.205916                0.092173                 0.017528   
min           0.000000                0.000000                 0.000000   
25%           0.479149                0.008913                 0.000000   
50%           0.660350                0.027465                 0.003225   
75%           0.789439                0.070330                 0.009427   
max           1.000000                0.889418                 0.796484   

       race_asian_rate  race_islander_rate  race_other_rate  \
count      8056.000000         8056.000000      8056.000000   
mean          0.132326            0.003787         0.128134   
std           0.151252            0.010553         0.133627   
min           0.000000            0.000000         0.000000   
25%           0.028888          

     ** It is shown in the above data distribution that the difference between the feature values has high proportional difference
     
     ** For example, average_household_income max value is 537130.22, however gini_index has max value of 0.73. 
     
     ** Since proportion difference leads our classifier for possible bias, normalization is crucial for the data

### B. Normalization

In [837]:
## Applying min max normalization to the dataset

data_norm=((dataset-dataset.min())/(dataset.max()-dataset.min()))
data_norm.head()

average_household_income  gini_index  per_capita_income  \
108                  0.150732    0.618723           0.106105   
109                  0.170568    0.550285           0.160917   
110                  0.768176    0.684935           0.687388   
111                  0.187196    0.734038           0.257556   
112                  0.217361    0.766727           0.368613   

     population_density  education_less_than_high_school_rate  \
108            0.065249                              0.047858   
109            0.051147                              0.102294   
110            0.015142                              0.064271   
111            0.063620                              0.066475   
112            0.074459                              0.026752   

     education_high_school_graduate_rate  education_college_rate  \
108                             0.430530                0.407565   
109                             0.389676                0.389957   
110                             0.286435                0.118164   
111                             0.516248                0.218769   
112                             0.229260                0.270535   

     education_bachelor_rate  education_master_rate  \
108                 0.146662               0.130838   
109                 0.162430               0.114693   
110                 0.289820               0.114970   
111                 0.251698               0.078411   
112                 0.340995               0.122080   

     education_professional_school_rate  education_doctoral_rate  \
108                            0.086712                 0.078799   
109                            0.057818                 0.048597   
110                            0.895199                 0.189207   
111                            0.247767                 0.205592   
112                            0.372157                 0.116767   

     race_white_rate  race_black_africa_rate  race_indian_alaska_rate  \
108         0.658069                0.015244                 0.004775   
109         0.652914                0.034470                 0.000000   
110         0.853829                0.011941                 0.000000   
111         0.814982                0.018515                 0.007094   
112         0.801053                0.005921                 0.003306   

     race_asian_rate  race_islander_rate  race_other_rate  race_two_more_rate  \
108         0.180700             0.01537         0.102679            0.053075   
109         0.136572             0.00000         0.143073            0.048637   
110         0.062435             0.00000         0.022079            0.056456   
111         0.111381             0.00000         0.025509            0.035680   
112         0.151220             0.00000         0.035693            0.017262   

     poverty_family_below_poverty_level_rate  heating_fuel_gas_rate  \
108                                 0.030888               0.788328   
109                                 0.060877               0.891746   
110                                 0.018315               0.801512   
111                                 0.147354               0.559985   
112                                 0.054127               0.704210   

     heating_fuel_electricity_rate  heating_fuel_fuel_oil_kerosene_rate  \
108                       0.179443                             0.000000   
109                       0.104195                             0.000000   
110                       0.186200                             0.000000   
111                       0.418578                             0.000000   
112                       0.270093                             0.014317   

     heating_fuel_coal_coke_rate  heating_fuel_solar_rate  \
108                          0.0                 0.150193   
109                          0.0                 0.000000   
110                          0.0                 0.162969   
111                          0.0

In [838]:
data_norm.describe()

average_household_income   gini_index  per_capita_income  \
count               8056.000000  8056.000000        8056.000000   
mean                   0.144660     0.567060           0.169697   
std                    0.082833     0.083271           0.103189   
min                    0.000000     0.000000           0.000000   
25%                    0.088346     0.511337           0.095875   
50%                    0.126384     0.561552           0.146361   
75%                    0.178792     0.613854           0.214631   
max                    1.000000     1.000000           1.000000   

       population_density  education_less_than_high_school_rate  \
count         8056.000000                           8056.000000   
mean             0.049617                              0.190524   
std              0.055680                              0.159264   
min              0.000000                              0.000000   
25%              0.015402                              0.062448   
50%              0.036438                              0.140160   
75%              0.063722                              0.283611   
max              1.000000                              1.000000   

       education_high_school_graduate_rate  education_college_rate  \
count                          8056.000000             8056.000000   
mean                              0.399291                0.294265   
std                               0.153796                0.090287   
min                               0.000000                0.000000   
25%                               0.292040                0.231739   
50%                               0.409216                0.295907   
75%                               0.510778                0.357706   
max                               1.000000                1.000000   

       education_bachelor_rate  education_master_rate  \
count              8056.000000            8056.000000   
mean                  0.193033               0.075635   
std                   0.113595               0.066832   
min                   0.000000               0.000000   
25%                   0.096597               0.023827   
50%                   0.180932               0.056847   
75%                   0.280578               0.111044   
max                   1.000000               1.000000   

       education_professional_school_rate  education_doctoral_rate  \
count                         8056.000000              8056.000000   
mean                             0.096285                 0.055221   
std                              0.116636                 0.080839   
min                              0.000000                 0.000000   
25%                              0.017493                 0.004085   
50%                              0.056986                 0.027343   
75%                              0.128706                 0.073857   
max                              1.000000                 1.000000   

       race_white_rate  race_black_africa_rate  race_indian_alaska_rate  \
count      8056.000000             8056.000000              8056.000000   
mean          0.623674                0.066963                 0.009757   
std           0.205916                0.103633                 0.022006   
min           0.000000                0.000000                 0.000000   
25%           0.479149                0.010021                 0.000000   
50%           0.660350                0.030880                 0.004049   
75%           0.789439                0.079074                 0.011835   
max           1.000000                1.000000                 1.000000   

       race_asian_rate  race_islander_rate  race_other_rate  \
count      8056.000000         8056.000000      8056.000000   
mean          0.144905            0.015305         0.128134   
std           0.165631            0.042652         0.133627   
min           0.000000            0.000000         0.000000   
25%           0.031634          

    ** The dataset is normalized using min-max normalization as shown and have uniform distribution among features

In [840]:
labels = data['solar_system_count'][ind].values
labels = np.where(labels > 0, 1, 0)
labels

array([1, 1, 1, ..., 0, 1, 0])

In [846]:
# split into training and validations subsets

from sklearn.model_selection import train_test_split

X_train, X_validation, y_train, y_validation = train_test_split(data_norm, labels, test_size = 0.4, random_state = 42, shuffle = True, stratify = labels)

In [847]:
# Split the validation data into validation and test data

X_val, X_test, y_val, y_test = train_test_split(X_validation, y_validation, test_size=0.5, random_state=42, shuffle=True ,stratify=y_validation )

In [853]:
def print_results(results):
    print("Best Params: {}\n".format(results.best_params_))   
    means = results.cv_results_['mean_test_score']
    stds = results.cv_results_['std_test_score']
    for mean, std, params in zip(means, stds, results.cv_results_['params']):
        print('{} (+/-{}) for {}'.format(round(mean,3), round(std*2, 3), params))

In [900]:
rf = RandomForestClassifier()
parameters = {
    'n_estimators': [250, 300, 350],
    'max_depth': [32,None],
    'max_features': [2, 3, 4],
    
    
}

cv = GridSearchCV(rf, parameters, cv=10, scoring='accuracy')
cv.fit(X_train, y_train)


GridSearchCV(cv=10, error_score=nan,
             estimator=RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                              class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features='auto',
                                              max_leaf_nodes=None,
                                              max_samples=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators=100, n_jobs=None,
                                              oob_score=False,
                                              rand

In [901]:
print_results(cv)

Best Params: {'max_depth': None, 'max_features': 4, 'n_estimators': 300}

0.971 (+/-0.006) for {'max_depth': 32, 'max_features': 2, 'n_estimators': 250}
0.971 (+/-0.006) for {'max_depth': 32, 'max_features': 2, 'n_estimators': 300}
0.971 (+/-0.005) for {'max_depth': 32, 'max_features': 2, 'n_estimators': 350}
0.972 (+/-0.007) for {'max_depth': 32, 'max_features': 3, 'n_estimators': 250}
0.972 (+/-0.007) for {'max_depth': 32, 'max_features': 3, 'n_estimators': 300}
0.971 (+/-0.007) for {'max_depth': 32, 'max_features': 3, 'n_estimators': 350}
0.972 (+/-0.006) for {'max_depth': 32, 'max_features': 4, 'n_estimators': 250}
0.971 (+/-0.006) for {'max_depth': 32, 'max_features': 4, 'n_estimators': 300}
0.972 (+/-0.006) for {'max_depth': 32, 'max_features': 4, 'n_estimators': 350}
0.971 (+/-0.006) for {'max_depth': None, 'max_features': 2, 'n_estimators': 250}
0.971 (+/-0.005) for {'max_depth': None, 'max_features': 2, 'n_estimators': 300}
0.971 (+/-0.007) for {'max_depth': None, 'max_feature

In [902]:
val_prediction = cv.predict(X_val)

In [903]:
print('Accuracy score for Testing Dataset = ', accuracy_score(y_val, val_prediction))

Accuracy score for Testing Dataset =  0.9745499689633768


In [904]:
test_prediction = cv.predict(X_test)

In [905]:
print('Accuracy score for Testing Dataset = ', accuracy_score(y_test, test_prediction))

Accuracy score for Testing Dataset =  0.9745657568238213


In [916]:
tot_predictions = cv.predict(data_norm)

In [917]:
print('Accuracy score for Testing Dataset = ', accuracy_score(labels, tot_predictions))

Accuracy score for Testing Dataset =  0.9895729890764647


In [937]:
class_data = data_norm.copy()
class_data.shape

(8056, 94)

In [938]:
class_data['prediction'] = tot_predictions
class_data = class_data[class_data['prediction'] == 1]
class_data.drop('prediction', axis=1)
class_data.shape

average_household_income  gini_index  per_capita_income  \
108                    0.150732    0.618723           0.106105   
109                    0.170568    0.550285           0.160917   
110                    0.768176    0.684935           0.687388   
111                    0.187196    0.734038           0.257556   
112                    0.217361    0.766727           0.368613   
...                         ...         ...                ...   
69082                  0.116513    0.441508           0.140303   
70866                  0.089078    0.666713           0.110814   
70867                  0.107993    0.611768           0.135334   
70868                  0.095292    0.596189           0.093281   
70871                  0.116646    0.726944           0.140314   

       population_density  education_less_than_high_school_rate  \
108              0.065249                              0.047858   
109              0.051147                              0.102294   
110              0.015142                              0.064271   
111              0.063620                              0.066475   
112              0.074459                              0.026752   
...                   ...                                   ...   
69082            0.000401                              0.101741   
70866            0.000558                              0.258131   
70867            0.000012                              0.153741   
70868            0.000363                              0.394062   
70871            0.004889                              0.235950   

       education_high_school_graduate_rate  education_college_rate  \
108                               0.430530                0.407565   
109                               0.389676                0.389957   
110                               0.286435                0.118164   
111                               0.516248                0.218769   
112                               0.229260                0.270535   
...                                    ...                     ...   
69082                             0.340486                0.550097   
70866                             0.595175                0.294500   
70867                             0.584390                0.352780   
70868                             0.469954                0.263158   
70871                             0.558203                0.279339   

       education_bachelor_rate  education_master_rate  \
108                   0.146662               0.130838   
109                   0.162430               0.114693   
110                   0.289820               0.114970   
111                   0.251698               0.078411   
112                   0.340995               0.122080   
...                        ...                    ...   
69082                 0.103288               0.052224   
70866                 0.119752               0.014193   
70867                 0.159918               0.020590   
70868                 0.076923               0.016194   
70871                 0.118388               0.042769   

       education_professional_school_rate  education_doctoral_rate  \
108                              0.086712                 0.078799   
109                              0.057818                 0.048597   
110                              0.895199                 0.189207   
111                              0.247767                 0.205592   
112                              0.372157                 0.116767   
...                                   ...                      ...   
69082                            0.062315                 0.000000   
70866                            0.007521                 0.002156   
70867                            0.014547                 0.015014   
70868                            0.009535                 0.006560   
70871                            0.116487                 0.015819   

       race_white_rate  race_black_africa_

(7879, 94)

In [ ]:
          ** We have classified data until this stage

In [932]:
reg_labels = data['solar_system_count'][class_data.index]
reg_labels

108       69.0
109      100.0
110       72.0
111       44.0
112       21.0
         ...  
69082     12.0
70866    346.0
70867      8.0
70868     53.0
70871    219.0
Name: solar_system_count, Length: 7879, dtype: float64

In [ ]:
class

In [921]:
data_norm.head()

average_household_income  gini_index  per_capita_income  \
108                  0.150732    0.618723           0.106105   
109                  0.170568    0.550285           0.160917   
110                  0.768176    0.684935           0.687388   
111                  0.187196    0.734038           0.257556   
112                  0.217361    0.766727           0.368613   

     population_density  education_less_than_high_school_rate  \
108            0.065249                              0.047858   
109            0.051147                              0.102294   
110            0.015142                              0.064271   
111            0.063620                              0.066475   
112            0.074459                              0.026752   

     education_high_school_graduate_rate  education_college_rate  \
108                             0.430530                0.407565   
109                             0.389676                0.389957   
110                             0.286435                0.118164   
111                             0.516248                0.218769   
112                             0.229260                0.270535   

     education_bachelor_rate  education_master_rate  \
108                 0.146662               0.130838   
109                 0.162430               0.114693   
110                 0.289820               0.114970   
111                 0.251698               0.078411   
112                 0.340995               0.122080   

     education_professional_school_rate  education_doctoral_rate  \
108                            0.086712                 0.078799   
109                            0.057818                 0.048597   
110                            0.895199                 0.189207   
111                            0.247767                 0.205592   
112                            0.372157                 0.116767   

     race_white_rate  race_black_africa_rate  race_indian_alaska_rate  \
108         0.658069                0.015244                 0.004775   
109         0.652914                0.034470                 0.000000   
110         0.853829                0.011941                 0.000000   
111         0.814982                0.018515                 0.007094   
112         0.801053                0.005921                 0.003306   

     race_asian_rate  race_islander_rate  race_other_rate  race_two_more_rate  \
108         0.180700             0.01537         0.102679            0.053075   
109         0.136572             0.00000         0.143073            0.048637   
110         0.062435             0.00000         0.022079            0.056456   
111         0.111381             0.00000         0.025509            0.035680   
112         0.151220             0.00000         0.035693            0.017262   

     poverty_family_below_poverty_level_rate  heating_fuel_gas_rate  \
108                                 0.030888               0.788328   
109                                 0.060877               0.891746   
110                                 0.018315               0.801512   
111                                 0.147354               0.559985   
112                                 0.054127               0.704210   

     heating_fuel_electricity_rate  heating_fuel_fuel_oil_kerosene_rate  \
108                       0.179443                             0.000000   
109                       0.104195                             0.000000   
110                       0.186200                             0.000000   
111                       0.418578                             0.000000   
112                       0.270093                             0.014317   

     heating_fuel_coal_coke_rate  heating_fuel_solar_rate  \
108                          0.0                 0.150193   
109                          0.0                 0.000000   
110                          0.0                 0.162969   
111                          0.0

In [866]:
def evaluate(classifier, training_instances, training_labels):
    metrics = cross_validate(classifier, training_instances, training_labels, cv=10, 
                             n_jobs=-1, scoring=['accuracy'])
    print("Accuracy: %0.4f (+/- %0.4f)" % (metrics['test_accuracy'].mean(), metrics['test_accuracy'].std() * 2))
    print("Mean fit time: %0.4f ms" % (metrics['fit_time'].mean()))

In [867]:
voting = VotingClassifier([ 
    ('knn', OneVsRestClassifier(KNeighborsClassifier(n_neighbors = 4))),
    ('et', OneVsRestClassifier(ExtraTreesClassifier(n_estimators = 1000))),
    ('xgboost', XGBClassifier(learning_rate =0.1, n_estimators=1000, num_class = 12,
                             min_child_weight=1, gamma=0,subsample=0.8,colsample_bytree=0.8,
                             objective= 'multi:softmax', nthread=4,scale_pos_weight=1,
                             seed=27, early_stopping_rounds=70, verbose=False)),
    ('randomforest', RandomForestClassifier(n_estimators=1000)),
    ('ada', AdaBoostClassifier(n_estimators=1000))], voting='soft', weights =[1,2,1,2,1])

In [868]:
voting.fit(X_train, y_train)

VotingClassifier(estimators=[('knn',
                              OneVsRestClassifier(estimator=KNeighborsClassifier(algorithm='auto',
                                                                                 leaf_size=30,
                                                                                 metric='minkowski',
                                                                                 metric_params=None,
                                                                                 n_jobs=None,
                                                                                 n_neighbors=4,
                                                                                 p=2,
                                                                                 weights='uniform'),
                                                  n_jobs=None)),
                             ('et',
                              OneVsRestClassifier(estimator=ExtraTreesClassifier(bootstra

In [891]:
val_prediction = voting.predict(X_val)

In [892]:
print('Accuracy score for Testing Dataset = ', accuracy_score(y_val, val_prediction))

Accuracy score for Testing Dataset =  0.9770328988206083


In [893]:
test_prediction = voting.predict(X_test)

In [894]:
print('Accuracy score for Testing Dataset = ', accuracy_score(y_test, test_prediction))

Accuracy score for Testing Dataset =  0.9727047146401985


In [880]:
svc = SVC()
parameters = { 'kernel': ['linear','rbf'], 'C': [0.1, 1, 10]}

cv = GridSearchCV(svc, parameters, cv=10)
cv.fit(X_train, y_train)

print_results(cv)

GridSearchCV(cv=10, error_score=nan,
             estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                           class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='scale', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='deprecated', n_jobs=None,
             param_grid={'C': [0.1, 1, 10], 'kernel': ['linear', 'rbf']},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

Best Params: {'C': 10, 'kernel': 'rbf'}

0.969 (+/-0.002) for {'C': 0.1, 'kernel': 'linear'}
0.969 (+/-0.002) for {'C': 0.1, 'kernel': 'rbf'}
0.969 (+/-0.002) for {'C': 1, 'kernel': 'linear'}
0.969 (+/-0.002) for {'C': 1, 'kernel': 'rbf'}
0.969 (+/-0.005) for {'C': 10, 'kernel': 'linear'}
0.971 (+/-0.005) for {'C': 10, 'kernel': 'rbf'}


In [885]:
val_prediction = cv.predict(X_val)

In [886]:
print('Accuracy score for Testing Dataset = ', accuracy_score(y_val, val_prediction))

Accuracy score for Testing Dataset =  0.9770328988206083


In [887]:
test_prediction = cv.predict(X_test)

In [888]:
print('Accuracy score for Testing Dataset = ', accuracy_score(y_test, test_prediction))

Accuracy score for Testing Dataset =  0.9727047146401985


In [895]:
mlp = MLPClassifier()
parameters = { 'hidden_layer_sizes': [(10,), (50,), (100,)],
                'activation': ['tanh', 'relu', 'linear'],
                 'learning_rate': ['constant', 'adaptive', 'invscaling']}

cv = GridSearchCV(mlp, parameters, cv=5)
cv.fit(X_train, y_train,)

print_results(cv)

GridSearchCV(cv=5, error_score=nan,
             estimator=MLPClassifier(activation='relu', alpha=0.0001,
                                     batch_size='auto', beta_1=0.9,
                                     beta_2=0.999, early_stopping=False,
                                     epsilon=1e-08, hidden_layer_sizes=(100,),
                                     learning_rate='constant',
                                     learning_rate_init=0.001, max_fun=15000,
                                     max_iter=200, momentum=0.9,
                                     n_iter_no_change=10,
                                     nesterovs_momentum=True, power_t=0.5,
                                     random_state...uffle=True,
                                     solver='adam', tol=0.0001,
                                     validation_fraction=0.1, verbose=False,
                                     warm_start=False),
             iid='deprecated', n_jobs=None,
             param_grid={'acti

Best Params: {'activation': 'relu', 'hidden_layer_sizes': (10,), 'learning_rate': 'invscaling'}

0.97 (+/-0.006) for {'activation': 'tanh', 'hidden_layer_sizes': (10,), 'learning_rate': 'constant'}
0.971 (+/-0.005) for {'activation': 'tanh', 'hidden_layer_sizes': (10,), 'learning_rate': 'adaptive'}
0.97 (+/-0.006) for {'activation': 'tanh', 'hidden_layer_sizes': (10,), 'learning_rate': 'invscaling'}
0.97 (+/-0.007) for {'activation': 'tanh', 'hidden_layer_sizes': (50,), 'learning_rate': 'constant'}
0.97 (+/-0.009) for {'activation': 'tanh', 'hidden_layer_sizes': (50,), 'learning_rate': 'adaptive'}
0.97 (+/-0.007) for {'activation': 'tanh', 'hidden_layer_sizes': (50,), 'learning_rate': 'invscaling'}
0.97 (+/-0.006) for {'activation': 'tanh', 'hidden_layer_sizes': (100,), 'learning_rate': 'constant'}
0.971 (+/-0.005) for {'activation': 'tanh', 'hidden_layer_sizes': (100,), 'learning_rate': 'adaptive'}
0.971 (+/-0.005) for {'activation': 'tanh', 'hidden_layer_sizes': (100,), 'learning_rat

In [896]:
val_prediction = cv.predict(X_val)

In [897]:
print('Accuracy score for Testing Dataset = ', accuracy_score(y_val, val_prediction))

Accuracy score for Testing Dataset =  0.9757914338919925


In [898]:
test_prediction = cv.predict(X_test)

In [899]:
print('Accuracy score for Testing Dataset = ', accuracy_score(y_test, test_prediction))

Accuracy score for Testing Dataset =  0.9714640198511166


In [776]:
ind = data_norm.index
cali_data['r'][ind]

108      10
109      10
110      10
111      10
112      10
         ..
70868    10
70869    10
70870    10
70871    10
71298    10
Name: rebate, Length: 8056, dtype: int64

In [790]:
data['solar_system_count'][ind].value_counts()

0.0       253
12.0      111
11.0      107
10.0      107
4.0       106
         ... 
529.0       1
490.0       1
1116.0      1
596.0       1
524.0       1
Name: solar_system_count, Length: 541, dtype: int64

In [789]:
data_norm.columns

Index(['average_household_income', 'gini_index', 'per_capita_income',
       'population_density', 'education_less_than_high_school_rate',
       'education_high_school_graduate_rate', 'education_college_rate',
       'education_bachelor_rate', 'education_master_rate',
       'education_professional_school_rate', 'education_doctoral_rate',
       'race_white_rate', 'race_black_africa_rate', 'race_indian_alaska_rate',
       'race_asian_rate', 'race_islander_rate', 'race_other_rate',
       'race_two_more_rate', 'poverty_family_below_poverty_level_rate',
       'heating_fuel_gas_rate', 'heating_fuel_electricity_rate',
       'heating_fuel_fuel_oil_kerosene_rate', 'heating_fuel_coal_coke_rate',
       'heating_fuel_solar_rate', 'heating_fuel_other_rate',
       'heating_fuel_none_rate', 'household_count', 'average_household_size',
       'housing_unit_count', 'housing_unit_occupied_count',
       'housing_unit_median_value', 'housing_unit_median_gross_rent',
       'age_18_24_rate', 'age